### DISCUSSION TITLE
Upstream status for touchscreen drivers

### DISCUSSION THREAD

**Discussion by InnovatumIT** [linux-surface/linux-surface#1305] [General]
Hi!

I'm just curious to know if there are any updates on the upstream status of Surface touchscreen drivers to the stock linux kernel?

The reason Im asking is that I use the, otherwise excelent, distro "Porteus Kiosk". Despite having multiple Surface pros (4, 5 and 6) I can't get touch to work on any of them. In contact with that project I was informed that they won't include custom drivers, only the stock kernel ones. I guess this is to avoid added complexity.

Is there any hope of upstream addions in the near future? :)

**Comment by StollD** [comment]
This topic is a bit difficult, because there are so many variations of the touchscreens, that we need three different drivers to support them all. IPTS for everything up to the Surface Pro 7+, ITHC for everything after that, and SPI-HID for the AMD based models. Lets exclude SPI-HID because nobody was able to make it work yet. That leaves us with IPTS and ITHC that we could theoretically upstream.

I cannot speak about ITHC because I don't know how it or the hardware it drives works. I only really know the older IPTS driver, because I wrote and maintain that. And for multiple reasons, upstreaming it is currently not on my agenda:

1. Just the thought of contributing to upstream linux gives me huge anxiety somehow. Its a combination of the email based patch submission process, combined with a lack of documentation about expected practices.  And I don't mean coding style, I mean things like how to properly document a hardware interface or how to design a userspace interface. The expected way to learn about these things is to have read the mailing list over the last 10 years, at least that's what it feels like. Combine this with the fact that I am working without a spec sheet and essentially from a badly written reference driver and guesswork and you have a situation where I have no idea if upstream would take the driver as-is or throw it out the door.
2. I cannot promise to maintain the driver. If my surface were to die tomorrow I couldn't promise you that I would buy another one. Even if I would buy another one, the newer models don't have the same hardware so I would have to buy an old model or a used one. And I don't want to throw code over the wall and run away from it. Upstreaming a driver for me is a promise to keep looking after it. 
3. Intels behaviour in this matter is - and I apologize for my choice of words - pissing me off. Whether it is touchscreens or cameras: Intel doesn't care about the (laptop-)hardware that they put out into the wild. If you are lucky enough you get a reference driver of dubious quality that the community can then at least use to learn something about the hardware. But then they go ahead and change the hardware every two generations for no apparent reason. For example lets take a look at the evolution of the surface touchscreens:
    * SP3: A simple SPI touchscreen. The calculation of coordinates from raw measurements is done in an embedded chipset, without the OS having to worry about it. Very easy to drive.
    * SP4: Intel Precise Touch & Stylus. The Intel ME acts as an interface to get raw touchscreen data, that is converted into coordinates using the GPU. Relied on modifications to the graphics drivers and support in the GPU firmware. Hard to drive, but it does benefit latency and flexibility.
    * SP7: Intel has now decided that they dont want to have this stuff in their GPU firmware and rip it out without a replacement. Instead of on the GPU, the conversion from raw measurements to coordinates is now happening on the CPU which is *very* inefficient. Additionally, the behaviour of the hardware changed to such a degree that it would warrant writing a different driver - except that it has the same device ID than the "old" IPTS, making two different drivers pretty much impossible. You now have to find the lowest common denominator in behaviour between the two, or special case everything. Great.
    * SP7+: One generation later Intel Touch Host Controller gets introduced. Essentially it does the same as IPTS, except that the hardware interface is not designed around the GPU anymore. But at this point, it is essentially just a HID transport driver. There is nothing touchscreen specific about it anymore. You could pass Mouse data or keyboard inputs or whatever you want - just like you could pass the touchscreen data over USB. Its complexity for the sake of being complex. Additionally, this time, they don't even release a reference driver, even though Microsoft is no longer the only OEM using this hardware.
    * SLS2: This is where Intel seems to loose their mind: Now the THC is gone, and it was replaced by Quick SPI which is ... an SPI port / controller / whatever. It creates SPI devices that Microsoft then uses with their SPI-HID protocol. And again they use the same device ID as the THC, which means forcing a driver to support both, ITHC and Quick SPI. Just imagine this: The SP4 was released in 2015. It took them 8 years to go from an SPI touchscreen to a ME/GPU accelerated touchscreen to a weird ME based touchscreen to a proper PCI based touchscreen to an SPI touchscreen.

Sorry for the "little" rant. To be clear, that last point about Intel is not the main reason why I am not thinking about upstreaming right now. The other two points are much more important. But it also doesn't really incentivize me to overcome them if I am honest. 

Another issue is that upstreaming the driver would only give you basic singletouch support. To get multitouch and pen support, distributions would have to package and integrate iptsd. Its not as big of a hurdle as the kernel, but its still a factor to consider.

**Reply by InnovatumIT** [reply]
Thank you for the thurough explanation of the situation!  I really appreciate all the hard work you and the other contributors put into this project.
